### DISCUSSION TITLE
Will node interpret typescript natively in the future?

### DISCUSSION THREAD

**Discussion by Pacheco95** [nodejs/help#5025] [Ideas]
Is there a plan for that or only in my dreams?

**Comment by Trott** [comment]
> Is there a plan for that

There is no plan for that, but that doesn't mean it will never happen.

**Comment by gitFoxCode** [comment]
@Trott This should be included in future versions of NodeJS. Every more advanced project requires typescript, and the fact that you have to install additional libraries for this is a terrible inconvenience. Especially since during the conference I got the impression that NodeJS is adding native features looking at the number of monthly downloads of a given library, then look at the number of people using typescript.

**Reply by Trott** [reply]
> This should be included in future versions of NodeJS.

If you want native TypeScript today, you can consider using `deno`. My understanding is that incorporating TypeScript natively for them has had upsides of course but also downsides (particularly around performance and maintenance), but I don't follow `deno` internal development that closely. Perhaps someone who has been involved in `deno` maintenance can comment. @bnoordhuis @cjihrig 

> NodeJS is adding native features looking at the number of monthly downloads of a given library

That is *a* metric but certainly not the only metric.

Long-term, it might be better (and perhaps even easier) to get the relevant/important parts of TypeScript into the ECMAScript spec and then implemented directly in V8 than to get Node.js to support TypeScript natively. I believe there is already a proposal for types. I'm personally concerned that TypeScript will go the way of CoffeeScript--wildly popular until the high-value stuff got implemented in ECMAScript and then people didn't really need it as much anymore. (No offense to the designers of CoffeeScript--in fact, quite the opposite. Nothing says "success" as much as "what you did was so valuable that ECMAScript adopted it natively".)

@nodejs/tsc Is there a sense among people on the TSC about Node.js native support of TypeScript in the future? Is it "this is never going to happen" or "eh, maybe, who knows" or "if you want this to happen, here is a list of 10 steps that will need to happen, but be forewarned that you might do a ton of work to get to step 4 and then get stuck there for literally years" or whatever?

**Reply by ronag** [reply]
I don't see this happening. I think the closest we are going to get is type annotations https://github.com/tc39/proposal-type-annotations.

**Reply by GeoffreyBooth** [reply]
> Is there a sense among people on the TSC about Node.js native support of TypeScript in the future?

The current plan is for this to be supported via the [Loaders API](https://nodejs.org/api/esm.html#loaders), and it currently is; the `ts-node` project comes pretty close to providing a seemingly built-in experience for TypeScript in Node. There are various UX improvements we’d like to make, like providing a way to register the loader via configuration (like in `package.json`) rather than a CLI flag, and TypeScript-specific errors like messages that link to a docs page to guide users on how to enable TypeScript support. None of these are “native” like Deno, but they’re hopefully close enough for most use cases. I don’t see us doing anything further anytime soon. Especially if https://github.com/tc39/proposal-type-annotations lands, TypeScript itself will become basically a linting tool and wouldn’t need to be supported in the runtime.

**Reply by mcollina** [reply]
The above was more or less agreed upon in https://github.com/nodejs/node/issues/43818.
